src/TestSuite/IntegrationTestTrait.php

```diff
@@ -511,6 +513,55 @@ protected function _sendRequest(array|string $url, string $method, array|string
         }
     }
 
+    /**
+     * Resolve the provided URL into a string.
+     *
+     * @param array|string $url The URL array/string to resolve.
+     * @return string
+     * @since 5.1.0
+     */
+    public function resolveUrl(array|string $url): string
+    {
+        // If we need to resolve a Route URL but there are no routes, load routes.
+        if (is_array($url) && count(Router::getRouteCollection()->routes()) === 0) {
+            return $this->resolveRoute($url);
+        }
+
+        return Router::url($url);
+    }
+
+    /**
+     * Convert a URL array into a string URL via routing.
+     *
+     * @param array $url The url to resolve
+     * @return string
+     * @since 5.1.0
+     */
+    protected function resolveRoute(array $url): string
+    {
+        $app = $this->createApp();
+
+        // Simulate application bootstrap and route loading.
+        // We need both to ensure plugins are loaded.
+        $app->bootstrap();
+        if ($app instanceof PluginApplicationInterface) {
+            $app->pluginBootstrap();
+        }
+        $builder = Router::createRouteBuilder('/');
+
+        if ($app instanceof RoutingApplicationInterface) {
+            $app->routes($builder);
+        }
+        if ($app instanceof PluginApplicationInterface) {
+            $app->pluginRoutes($builder);
+        }
+
+        $out = Router::url($url);
+        Router::resetRoutes();
```

Comment:
This is a clever solution.

Reply:
I always wondered how many people actually use a custom application class which doesn't implement those interfaces for Plugins and Routing... feels to me like unnecessary decoupling things.

Reply:
@LordSimal This is mainly needed for integration tests in the core core rather than app's integration tests.

Reply:
Well then this comes down to the old topic of mocks imho. When we upgraded to PHPUnit 10 we talked about the fact, that PHPUnit discourages mocks more and more over time.

I know we decided, that all new mocks should be done with Mockery instead of PHPUnit (if needed at all) but in this specific situation I can only see 3 places where a "pure" Application object is created:
* [RunnerTest](https://github.com/cakephp/cakephp/blob/6a1fe8e590194a79246e6f74c262012dfc1af5d4/tests/TestCase/Http/RunnerTest.php#L138C20-L138C31)
* [ErrorHandlerMiddlewareTest](https://github.com/cakephp/cakephp/blob/6a1fe8e590194a79246e6f74c262012dfc1af5d4/tests/TestCase/Error/Middleware/ErrorHandlerMiddlewareTest.php#L394)
* [RoutingMiddlewareTest](https://github.com/cakephp/cakephp/blob/6a1fe8e590194a79246e6f74c262012dfc1af5d4/tests/TestCase/Routing/Middleware/RoutingMiddlewareTest.php#L183)

All other test either already use a PHPUnit mock or the IntegrationTestTrait methods which do `$this->createApp()` and therefore just use the base App class (for users at least).

I'm just saying this could maybe be something that could be cleaned up for a maybe 6.0 release if there is nothing more than just a few core tests which benefit from this abstraction.